framework/src/loops/ComputeUserObjectsThread.C

```diff
@@ -185,6 +187,45 @@ ComputeUserObjectsThread::onInternalSide(const Elem * elem, unsigned int side)
       uo->execute();
 }
 
+void
+ComputeUserObjectsThread::onInterface(const Elem * elem, unsigned int side, BoundaryID bnd_id)
+{
+  std::cout << "ComputeUserObjectsThread::onInterface" << std::endl;
+  // Pointer to the neighbor we are currently working on.
+  const Elem * neighbor = elem->neighbor_ptr(side);
+
+  // Get the global id of the element and the neighbor
+  const dof_id_type elem_id = elem->id(), neighbor_id = neighbor->id();
+
+  std::vector<UserObject *> userobjs;
+  queryBoundary(Interfaces::InterfaceUserObject, bnd_id, userobjs);
+  if (userobjs.size() == 0)
+    return;
+
+  if (!((neighbor->active() && (neighbor->level() == elem->level()) && (elem_id < neighbor_id)) ||
```

Comment:
What is the purpose of the test `elem_id < neighbor_id`?

Reply:
This ensures that you don't compute the object twice when doing the element loop...say first you compute element 0 with neighbor element 1 on element 0's zeroth side. When you come around and compute for element 1, you don't want to compute the side again when the neighbor is element 0.

Reply:
The reason I'm asking is that I checked this branch out and tried it on a 3x3 subdomain problem with an interface around the center subdomain and it looks like `execute()` only loops over half of the interface while it loops over the whole interface as I'd expect if I remove the  condition `elem_id < neighbor_id`.

Reply:
Interesting. If you make a branch with the bad behavior checked in, I will look at this when I get a chance. Or maybe @arovinelli would also check it out

Reply:
@vincentlaboure  This is interesting and is a case I didn't check. So we are expecting the object to be executed 6 times, once for each face of the cube but you are saying it is executed only 3 times? If this is the case, as @lindsayad  mentioned, please make a branch for it, I will then check what's going on.

Just for completness, @vincentlaboure have you checked if this is also the case for a boudnary using a sideUserObject?

Reply:
This logic is probably only for internal sides where we want to visit them once. It is perfectly fine to visit the same side on two sides depending on how users set up the side sets.

Reply:
@arovinelli I added two commits to your branch which I pushed to my [repo](https://github.com/vincentlaboure/moose/commits/boundaryRestrictableInternalSideUO).

The first commit adds an input file, `test/tests/userobjects/interface_user_object/interface_user_object_QP_3x3.i`. 

Without the second commit, only the top and right pieces of the interface (`y=2` and `x=2`, respectively) are looped over.
With the second commit, the entire interface is looped over.

Reply:
@vincentlaboure First of all thanks for spotting this and taking the time to make an example.
The idea is that one should loop over every face only once, because there is no reason to loop twice. However it seems that rather than doing that, it loops over only half of the sides... this is very strange. I'll look at it, and get back to you as soon as I have something new!

Reply:
@arovinelli You're welcome! I need your `InterfaceUserObject` for my own work so I'm looking forward to having it merged into MOOSE.

Reply:
You may want to take a look at https://github.com/idaholab/moose/blob/next/framework/src/loops/ComputeResidualThread.C#L157, where we do not have this. Even on the same side, if we create multiple side sets covering the side, we will run `onInterface` multiple times (which is desired), https://github.com/idaholab/moose/blob/next/framework/include/loops/ThreadedElementLoopBase.h#L229.

Reply:
@vincentlaboure I checked and you are right this is happening. So i'll fix it and I'll get back witha PR soon.
@YaqiWang maybe I misinterpreted the condition from the `onInternalSide`, which is the one I used to develop the `interfaceUO`.
In the internal side you have
```
  if (!((neighbor->active() && (neighbor->level() == elem->level()) && (elem_id < neighbor_id)) ||
        (neighbor->level() < elem->level())))
```
of which all the `->level` call should be for automatic mesh refinement  so it should be there.
I'll just take the `&& (elem_id < neighbor_id)` away.
As even with this away doesn't loop on the interface twice.

Reply:
That logic says, if  the neighbor element is active and on the same level (connecting to the current element, which is active, with the full side) and neighbor element has not been visited, or the neighbor has lower level (which means if the current element is that neighbor, its neighbor is not active, so the first condition does not satisfy), we do evaluations.

I think you should remove the if statement completely. When we refine an element, if a side is in a side set, probably the sub sides of that side after refinement are also pushed into the same side set. @lindsayad or any libMesh developers can possibly confirm this for us. This `onInternalSide` is just used to visit internal sides in any side sets. The same geometrical side can be visited multiple times if there multiple side sets including it. A side set may include the side on the two sides with respect to to master elements. It is noted that a side in side set is always defined with an element and an local side ID.